### PR TITLE
Improve nightly and beta CI

### PR DESCRIPTION
**Pull request by robertbastian**
* Decoupling `diplomat-coverage` from pinned nightly
  * It is written for a specific nightly version and we can run with that for the time being
* Decoupling `test-c-tiny` from pinned nightly
  * It requires using a toolchain with an old LLVM version
  * Updated to LLVM 16
* Run full CI in forced-nightly mode
  * We really want to know if FFI builds break with a new nightly 
  * This lets us easily bump pinned nightly because we know for each nightly if it works
* Call out beta instead of nightly sizes in docs
  * This is more useful for users, as beta will become stable, whereas anything can happen to nightly

### CODE REVIEW

File: components/datetime/src/helpers.rs
Comment (sffc) [inline components/datetime/src/helpers.rs:49]: Issue: Yes we care about size on nightly

File: components/datetime/src/helpers.rs
Comment (sffc) [inline components/datetime/src/helpers.rs:21]: I'd like these tests to run on latest nightly so that we can learn immediately about regressions or improvements.

**Review by sffc — CHANGES_REQUESTED** [review]

File: components/datetime/src/helpers.rs
Comment (sffc) [inline components/datetime/src/helpers.rs:48]:
```suggestion
                Some("beta") | Some("nightly") => size == $beta,
                Some("pinned-stable") => size == $pinned,
                _ => matches!(size, $pinned | $beta),
```

File: components/datetime/src/helpers.rs
Comment (sffc) [inline components/datetime/src/helpers.rs:35]: Nightly, beta, and stable could all be different sizes. I believe it was recently that they actually were all different at the same time. Right now this test only looks at stable and nightly. If we want it to also look at beta, we should change the macro to allow a beta size and a nightly size.

Comment (robertbastian) [reply components/datetime/src/helpers.rs:35]: Yes but that would give us 4 different combinations which I don't all want to copy paste right now. Can we cross that bridge when we get there?

Comment (sffc) [reply components/datetime/src/helpers.rs:35]: Just make two cases, all the same and all explicit.

Comment (sffc) [reply components/datetime/src/helpers.rs:35]: (this is what we currently have: just make the all-different case have three instead of two different values)

File: components/datetime/src/helpers.rs
Comment (sffc) [inline components/datetime/src/helpers.rs:21]: Nit: update the `ignore`d docs test

**Review by sffc — COMMENTED** [review]
Okay in the datetime size test but wait for @Manishearth approval

**Review by Manishearth — APPROVED** [review]

**Review by sffc — APPROVED** [review]

**Pull request merged**
